doc/PARAMETERS.md
## List of parameters

Here we provide an explanation of the various parameters of the model (passed though a parameter file, e.g. `parameters.txt`), as well as their default value when no parameter file is supplied, and their valid ranges. For more information, please refer to the source code or the scientific manuscript accompanying this project (link will be made available once published).

Note that hereafter, "positive" means ***including zero***, while "strictly positive" means ***excluding zero***. Also note that missing parameters in the parameter file will take their default value.

| Parameter name | Default value(s) | Accepted values | Description | Notes |
|--|--|--|--|--|
| `popsize` | `10` | Strictly positive integers | Number of individuals in the population
| `xstart` | `0` | Decimals | Initial trait value of all individuals in the population
| `tradeoff` | `1` | Positive decimals | Strength of the utilization trade-off between the two resources | A value of 0 means no trade-off |
| `alpha` | `0` | Decimals from zero to one | Degree to which resource encounter is determined by their abundance | A value of 0 means that the probability of choosing a resource when choosing at random (`beta` is 0) is proportional to the abundance of that resource in the habitat, wherease a value of 1 means that the probability of choosing a resource at random is 0.5 |
| `beta` | `1` | Decimals from zero to one  | Degree of optimal choice | A value of 0 means that individuals choose randomly, while a value of 1 means that individuals always choose the most advantageous resource |
| `delta` | `1` | Positive decimals | Resource discovery rate | The per capita amount of resources discovered by an individual with unit resource consumption rate alone in its habitat (i.e. initial slope of the resource discovery function) |
| `hsymmetry` | `1` | Decimals from zero to one | Habitat symmetry | A value of 0 means that the two resources are completely asymmetrically distributed among the two habitats, while a value of 1 means that they are completely symmetrically distributed |
| `nrounds` | `10` | Strictly positive integers | Number of feeding rounds in a feeding season |
| `mutrate` | `0.0001` | Decimals from zero to one | Per individual mutation rate | The probability of any given offspring to mutate from its parent |
| `mutsdev` | `0.02` | Positive decimals | Mutational standard deviation | Standard deviation of the normal distribution from which mutations are sampled |
| `dispersal` | `0.01` | Decimals from zero to one | Per capita probability of any individual to disperse to the opposite habitat during the dispersal step | A value of 0 means no dispersal |
| `tend` | `100` | Strictly positive integers | Number of time steps in a simulation | The simulation will run for `tend` time steps, starting at time step 0 |
| `tsave` | `20` | Strictly positive integers | Frequency of data saving (in time steps) | The output will be saved every `tsave` time steps, including the first and last ones, for a total of **`tend` / `tsave` + 1** time steps saved |
| `seed` | Clock-generated | Positive integers | Seed of the pseudo-random number generator | The clock is used to generate a pseudo-random seed. Make sure to set `savepars` to 1 to be able to retrieve the generated seed and reproduce a given simulation. |
| `savepars` | `0` | One or zero | Whether or not to save the parameters into a parameter log file called `paramlog.txt` | If set to 1, the parameters will be saved in a file called `paramlog.txt` in the working directory |
| `savedat` | `0` | One or zero | Whether or not to save output data | If set to 1, the output data will be saved in the requested output file(s) in the working directory. Check [here](OUTPUT.md) for details on how to choose which variables to save. |
| `choose` | `0` | One or zero | Whether or not to choose which output variables to save by providing a `whattosave.txt` file | If set to 1, the program will read the `whattosave.txt` file in the working directory to determine which variables to save. See [here](OUTPUT.md) for how this works |
| `verbose` | `0` | One or zero | Whether or not to display progress at each time step to the screen | If set to 1, the program will display the current time step and the number of individuals in the population at each time step |
| `memsave` | `1` | Decimals greater than the size of a double precision floating point number (usually 8 bytes, i.e. 8e-06 MB, on a 64 bit system), in MB | Memory that can be filled in by each data saving buffer (in MB) before writing to file | This is the amount of memory that can be used by each output file before it is written to disk. There will be as many open buffers as there are output files to save (see details [here](OUTPUT.md)). |

docs/PARAMETERS.md
## List of parameters

| Name | Meaning | Default | Note |
|--|--|--|--|
| popsize | Population size | 1000 | The population size is fixed through time |
| xstart | Initial trait value | -1 | The trait value of all individuals at the start of the simulation |
| tradeoff | Trade-off strength | 1 | The strength of the utilization trade-off between the two resources |
| beta | Degree of optimal choice | 0 | The probability with which individuals choose the most advantageous resource |
| delta | Resource discovery rate | 1 | A parameter scaling the degree of soft versus hard selection |
| resource | Quantity of resource | 100 | Both resources, in arbitrary units (replenished after each feeding round) |
| hsymmetry | Habitat symmetry | 1 | The degree of symmetry in resource distribution between the two habitats |
| nrounds | Number of feeding rounds | 10 | Number of feeding rounds in a feeding season |
| typeII | Type II resource choice (0/1) | 0 | Whether type II resource choice should be used instead of type I |
| mutrate | Mutation rate | 0.0001 | Per individual |
| mutsdev | Mutational standard deviation | 0.02 | Standard deviation of the normal distribution mutations are sampled from |
| dispersal | Dispersal rate | 0.1 | Chances that an individual disperses to the opposite habitat |
| tend | Simulation time | 1000 | Number of time steps in a simulation |
| tsave | Saving frequency | 10 | Save output every how many time steps? |
| seed | Seed for the random number generator | Determined by the clock | |
| savepars | Whether to save parameters (0/1) | 1 | If yes, saved to `paramlog.txt` |
| savelog | Whether to save screen output (0/1) | 0 | If yes, saved to `screenlog.txt` |
| talkative | Whether to output time steps to screen (0/1) | 1 | |
| whattosave | Which variable(s) to save as output | 1 | Click [here](OUTPUT.md) for details |


src/parameters.cpp
// This script contains all the functions of the Parameters structure.

#include "parameters.hpp"

// Create a default seed based on clock
size_t clockseed() {

    // Use the current time to generate a seed
    auto seed = std::chrono::high_resolution_clock::now().time_since_epoch().count();

    // Convert the seed to integer and return
    return static_cast<size_t>(seed);
}

Parameters::Parameters(const std::string &filename) :
    popsize(10u),
    xstart(0.0),
    tradeoff(1.0),
    alpha(0.0),
    beta(1.0),
    delta(1.0),
    hsymmetry(1.0),
    nrounds(10u),
    mutrate(0.0001),
    mutsdev(0.02),
    dispersal(0.1),
    tend(100u),
    tsave(20u),
    seed(clockseed()),
    savepars(true),
    savedat(false),
    verbose(false),
    choose(false),
    memsave(1.0)
{
    
    // filename: optional parameter input file

    // Read from file if needed
    if (filename != "") read(filename);

}

// Check that the parameter values are valid
void Parameters::check() const
{
    std::string msg = "No error detected";

    // Check the validity of the supplied values
    assert(popsize > 0u);
    assert(tradeoff >= 0.0);
    assert(alpha >= 0.0 && alpha <= 1.0);
    assert(beta >= 0.0 && beta <= 1.0);
    assert(delta >= 0.0);
    assert(hsymmetry >= 0.0 && hsymmetry <= 1.0);
    assert(nrounds > 0u);
    assert(mutrate >= 0.0 && mutrate <= 1.0);
    assert(mutsdev >= 0.0);
    assert(dispersal >= 0.0 && dispersal <= 1.0);
    assert(tend > 0u);
    assert(tsave > 0u);
    assert(memsave >= 0.0);
    
}

// Read parameters from a file
void Parameters::read(const std::string &filename)
{

    // filename: name of the file to read parameters from

    // Create a reader
    Reader reader(filename);

    // Open it
    reader.open();

    // For each line in the file...
    while (!reader.iseof()) {

        // Read a line
        reader.readline();

        // Skip empty line
        if (reader.isempty()) continue;

        // Skip if comment line
        if (reader.iscomment()) continue;

        // Check
        assert(!reader.isempty());
        assert(!reader.iscomment());
        
        // Current parameter name 
        std::string name = reader.getname();

        // Read the parameter value(s)
        if (name == "popsize") reader.readvalue<size_t>(popsize, chk::strictpos<size_t>);
        else if (name == "xstart") reader.readvalue<double>(xstart);
        else if (name == "tradeoff") reader.readvalue<double>(tradeoff, chk::positive<double>);
        else if (name == "alpha") reader.readvalue<double>(alpha, chk::proportion<double>);
        else if (name == "beta") reader.readvalue<double>(beta, chk::proportion<double>);
        else if (name == "delta") reader.readvalue<double>(delta, chk::positive<double>);
        else if (name == "hsymmetry") reader.readvalue<double>(hsymmetry, chk::proportion<double>);
        else if (name == "nrounds") reader.readvalue<size_t>(nrounds, chk::strictpos<size_t>);
        else if (name == "mutrate") reader.readvalue<double>(mutrate, chk::proportion<double>);
        else if (name == "mutsdev") reader.readvalue<double>(mutsdev, chk::positive<double>);
        else if (name == "dispersal") reader.readvalue<double>(dispersal, chk::proportion<double>);
        else if (name == "tend") reader.readvalue<size_t>(tend, chk::strictpos<size_t>);
        else if (name == "tsave") reader.readvalue<size_t>(tsave, chk::strictpos<size_t>);
        else if (name == "seed") reader.readvalue<size_t>(seed);
        else if (name == "savepars") reader.readvalue<bool>(savepars);
        else if (name == "savedat") reader.readvalue<bool>(savedat);
        else if (name == "verbose") reader.readvalue<bool>(verbose);
        else if (name == "choose") reader.readvalue<bool>(choose);
        else if (name == "memsave") reader.readvalue<double>(memsave, chk::enoughmb<double>);
        else
            reader.readerror();

    }

    // Close the file
    reader.close();

    // Check
    check();

}

// Save parameters to a file
void Parameters::save(const std::string &filename) const
{

    // filename: name of the file to save into

    // Create output file stream
    std::ofstream file(filename);
    
    // Check if the file is open
    if (!file.is_open()) 
        throw std::runtime_error("Unable to open file " + filename);

    // Write parameters to the file
    file << "popsize " << popsize << '\n';
    file << "xstart " << xstart << '\n';
    file << "tradeoff " << tradeoff << '\n';
    file << "alpha " << alpha << '\n';
    file << "beta " << beta << '\n';
    file << "delta " << delta << '\n';
    file << "hsymmetry " << hsymmetry << '\n';
    file << "nrounds " << nrounds << '\n';
    file << "mutrate " << mutrate << '\n';
    file << "mutsdev " << mutsdev << '\n';
    file << "dispersal " << dispersal << '\n';
    file << "tend " << tend << '\n';
    file << "tsave " << tsave << '\n';
    file << "seed " << seed << '\n';
    file << "savepars " << savepars << '\n';
    file << "savedat " << savedat << '\n';
    file << "verbose " << verbose << '\n';
    file << "choose " << choose << '\n';
    file << "memsave " << memsave << '\n';

    // Close the file
    file.close();

}